Add `rehype-mdx-toc` to export toc into MDX module

boning-w opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]).
  Leave the comments as they are: they do not show on GitHub.

  Please try to limit the scope,
  provide a general description of the changes,
  and remember it’s up to you to convince us to land it.

  We are excited about pull requests.
  Thank you!
-->

### Initial checklist

* [x] I read the support docs <!-- https://github.com/rehypejs/.github/blob/main/support.md -->
* [x] I read the contributing guide <!-- https://github.com/rehypejs/.github/blob/main/contributing.md -->
* [x] I agree to follow the code of conduct <!-- https://github.com/rehypejs/.github/blob/main/code-of-conduct.md -->
* [x] I searched issues and discussions and couldn’t find anything or linked relevant results below <!-- https://github.com/search?q=user%3Arehypejs&type=issues and https://github.com/orgs/rehypejs/discussions -->
* [x] I made sure the docs are up to date
* [x] I included tests (or that’s not needed)

### Description of changes

I wrote a rehype plugin [rehype-mdx-toc](https://github.com/boning-w/rehype-mdx-toc) that allows users to export the table of contents (TOC) data as a named export from an MDX module.

#### Why another table of contents plugin?

Since MDX supports exporting data from MDX modules, it's much more convinent to export the TOC — so that users can use it whatever they like, just like with [remark-mdx-frontmatter](https://github.com/remcohaszing/remark-mdx-frontmatter/tree/main).

Simple usage:

```js
import { compile, run } from "@mdx-js/mdx";
import rehypeMdxToc from "rehype-mdx-toc";
import * as runtime from "react/jsx-runtime";

const mdxContent = `
# Heading 1
## Heading 2
### Heading 3
`;

const compiled = await compile(mdxContent, {
  outputFormat: "function-body",
  rehypePlugins: [rehypeMdxToc],
});
const result = await run(compiled, { ...runtime });
const toc = result.toc;
console.log(toc);
```

With some bundlers — like [@mdx-js/esbuild](https://mdxjs.com/packages/esbuild/), [@mdx-js/rollup](https://mdxjs.com/packages/rollup/), and [@mdx-js/loader](https://mdxjs.com/packages/loader/), we can easily import the TOC data at build time:

```js
import MDXContent, { toc } from "./you-mdx-file.mdx";
```

<!--do not edit: pr-->

boning-w committed:
feat: add `rehype-mdx-toc` plugin

Signed-off-by: boning-w <wang.boning@outlook.com>

boning-w committed:
chore: change the position

Signed-off-by: boning-w <wang.boning@outlook.com>

ChristianMurphy reviewed (approved):
Thanks for sharing @boning-w!

Some suggestions:
* Consider widening dependencies to `x.0.0` to allow npm (and packageme manager friends) to de-duplicate the dependencies as much as possible https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/package.json#L47-L51
* What is the intent of `default` the option? Could it be made more specific than `unknown`? https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/src/index.ts#L38
* It is recommended to have a lint checker, whether that be prettier, xo, standard, or a custom ESLint config https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/package.json#L43
* These options should not be needed in `nodenext` module mode https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/tsconfig.json#L23-L24
* avoid skipping libcheck, it can hide upstream issues that may impact your users https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/tsconfig.json#L38

boning-w commented:
Hi @ChristianMurphy, I really appreciate for your code review.

- I have updated the dependencies versions to 'x.0.0'.
- I added Biome as a lint checker, and added it in the test script:
  `"test": "npx @biomejs/biome check ./src ./test && node --test"`.
- I removed the unnecessary options from `tsconfig.json` and also remove `skipLibCheck`.

Regarding:

> * What is the intent of `default` the option? Could it be made more specific than `unknown`? https://github.com/boning-w/rehype-mdx-toc/blob/f1cc7f1ad5a8573bfcd87ca82fc2133f5c6f17a6/src/index.ts#L38

The `default` option is to provide a fallback value for users when there are no headings in Markdown contents. By default. it is `undefine`, but maybe users want to configure their own default value when no headings are found. Since the value could be `undefine`, `[]` or others, I use the type `unknown`.

I commited these changes and published a new version `v1.1.0` to npm.

[Here is the new version](https://www.npmjs.com/package/rehype-mdx-toc).

remcohaszing commented:
Thanks! Your plugin looks really cool!

I just don’t think this is the right place to list it. None of the MDX specific plugins are listed in the rehype or remark plugin lists. A better place would be https://mdxjs.com/docs/extending-mdx/#list-of-plugins.

boning-w commented:
Got it! I may create a pull request there. Should I cloes this pull request?

remcohaszing commented:
Let’s have another maintainer close it if they agree with me. :)

JounQin commented:
I agree it should be listed as a mdx specified plugin instead.

wooorm commented:
I’ll close this here. Yes, please do add it there! Thanks everyone!
